scipy/integrate/tests/test_quadpack.py

```diff
@@ -503,6 +503,27 @@ def f(x, y, z):
             error_tolerance=6e-8
         )
 
+    def test_complex(self):
+        def tfunc(x):
+            return np.exp(1j*x)
+
+        assert_quad(quad(tfunc, 0, np.pi/2, complex_func=True),
+                    1+1j, error_tolerance=1e-6)
+
+        kwargs = {'a': 0, 'b': np.inf, 'full_output':True, 'weight':'cos', 'wvar': 1}
+        res_c = quad(tfunc, complex_func=True, **kwargs)
+        res_r = quad(np.cos, complex_func=False, **kwargs)
+        res_i = quad(np.sin, complex_func=False, **kwargs)
+
+        assert len(res_c[2]['real']) == len(res_r[2:])
+        assert res_c[2]['real'][2] == res_r[4]
+        assert res_c[2]['real'][1] == res_r[3]
+        assert np.allclose(res_c[2]['real'][0]['rslst'], res_r[2]['rslst'])
+        assert np.allclose(res_c[2]['real'][0]['erlst'], res_r[2]['erlst'])
+        assert np.allclose(res_c[2]['real'][0]['ierlst'], res_r[2]['ierlst'])
+
+        assert len(res_c[2]['imag']) == len(res_i[2:])
```

Comment:
1.
You should be able to do something like this
```python3
assert_equal(res_c[0:2], res_r[0:2] + res_i[0:2]*1j)
# or assert_allclose(res_c[0:2], res_r[0:2] + res_i[0:2]*1j, rtol=1e-15)
```
if `assert_equal` doesn't pass, use the `assert_allclose` version. TBH, it makes sense to me that exact equality might not pass because `np.exp(1j*x)` might not be defined exactly as `np.cos(x) + 1j * np.sin(x)`. 

2.
For element 2, just do the length check like
```python3
assert len(res_c[2]['real']) == len(res_r[2:]) == 3
assert len(res_c[2]['imag']) == len(res_r[2:]) == 1
```
Adding the expected number of element at the end makes it obvious that we're checking a case in which `message` and `explain` are present. 
I don't think we can expect to get equality of the dictionaries themselves. Explain in a comment that they are just not reliable. 

3. You should also be able to do
```python3
assert_equal(res_c[2]['real'][1:], res_r[3:])
```

Reply:
@cowden I see that some of this is resolved, but did you try `assert_equal` for number 1 above? If you are going to use `assert_allclose`, tighten the tolerance to show that they are essentially equal, and why not test `0` before `1`?

Also, in number 2, I was suggesting that you should just change the lengths, not all the separate elements. What if one of those checks in lines 536-540 or 546-551 starts failing some day? Is it more likely to be a real issue or a false alarm?

And what about number 3? Did that not work?

Reply:
Lines 536-540 and 546-551 are only testing an internal state.  I don't think it makes any more of a thorough test than what we already have by testing the results (integral and error).  I will remove those  lines.

Lines 532-533 do what you suggest in number 3 above.  Consolidating the test into a single line does work, but I separating the two comparisons makes identifying a failure more explicit.

Reply:
> separating the two comparisons makes identifying a failure more explicit.

Not convinced that we need that, but I can live with it. Thanks for making those changes!